federation/pkg/federation-controller/sync/controller.go

```diff
@@ -452,22 +486,23 @@ func syncToClusters(clustersAccessor clustersAccessorFunc, operationsAccessor op
 		return statusError
 	}
 
-	// Evertyhing is in order but let's be double sure
+	// Everything is in order but let's be double sure
 	return statusNeedsRecheck
 }
 
 type clusterObjectAccessorFunc func(clusterName string) (interface{}, bool, error)
 type clusterSelectorFunc func(map[string]string, map[string]string) (bool, error)
 
 // clusterOperations returns the list of operations needed to synchronize the state of the given object to the provided clusters
-func clusterOperations(adapter federatedtypes.FederatedTypeAdapter, clusters []*federationapi.Cluster, obj pkgruntime.Object, key string, accessor clusterObjectAccessorFunc, selector clusterSelectorFunc) ([]util.FederatedOperation, error) {
-	// The data should not be modified.
-	desiredObj := adapter.Copy(obj)
-	objMeta := adapter.ObjectMeta(desiredObj)
+func clusterOperations(adapter federatedtypes.FederatedTypeAdapter, clusters []*federationapi.Cluster, obj pkgruntime.Object, key string, userInfo interface{}, accessor clusterObjectAccessorFunc, selector clusterSelectorFunc) ([]util.FederatedOperation, error) {
 	kind := adapter.Kind()
 	operations := make([]util.FederatedOperation, 0)
 
 	for _, cluster := range clusters {
+		// The data should not be modified.
+		desiredObj := adapter.Copy(obj)
+		objMeta := adapter.ObjectMeta(desiredObj)
+
 		clusterObj, found, err := accessor(cluster.Name)
```

Comment:
While in the process of rebasing my work on this, I met with another problem...
In my scheduling algorithm, I need to make decisions as to which cluster gets an object (even if its not there), which cluster gets the object updated, and also remove object from some cluster if needed (even if it exists).
In this flow, the same cluster objects gets accessed 2 different places (at one place in PrepareForUpdate()using informer directly, and another one here...) If I try to keep this same flow, and try to use just accessed local object(at this line of comment), there is a possible race condition for me..(i get an object here, which wasnt present when I accessed in prepareforupdate, my earlier schedule decision is wrong). On the other hand if I stick to the schedule() computation made in PrepareForUpdate() (which tells me ab all three conditions - to remove object, to add object, to update object) already and ignore the object accessed here, I need to introduce some crooked logic only for this hpa adapter, which doesn't sound right to me......

So my need and use case still is the same, and kind of the super set of what is needed for replicasets (the rs controller never deletes an object from any cluster in reconcile) and the simple steps that I want to achieve are still same.
- get the list of clusters with the objects (with nil or some-other way to indicate if no object in that cluster).
- schedule makes decisions where to add, where to remove and where to update, conveys the result back.
- clusteroperations makes a list of operations to act on that decision.

Is there some way, we can update this logic to achieve above.

Additional note to this, which somehow I could not express properly in our discussion yesterday, was about the methods. Actually its a wrong term we are using. PrepareForUpdate( ) is an interface first (then a method on a structure), and our goal should be to keep the interfaces to minimum, and I sort of want to stress my opinion of keeping these interfaces to minimum. Our goal should also be to keep the information exchange to and from the interfaces to minimum (userstate here) especially if its not needed for any controller back in reconcile as of now (YAGNI). Implementers can choose to structure/break their code in as many methods/functions internally for the interface implementation. 

With that said, I am totally ready to rebase on top of this, if everybody thinks this is the right approach.
I just want this to reach to some conclusion, ASAP.. :-)
( I wrote all this because, I tried to rebase on top, this second time, and got stuck with no direct solution.. :-( )